LLVM 3.9

cocreature opened the pull request:
LLVM 3.9 is supposed to be out next week, I created this using rc1. I still need to cleanup some stuff (which is why it is marked WIP) but all tests are passing.

I haven’t yet figured out why travis is failing, maybe the version in the repo is not up2date. Also this should go in a new branch but I have no idea how I can make a pullrequest that is supposed to create a new branch.

While it is probably best to wait to merge this until 3.9 is released, I wanted to create this PR now to avoid duplicating work.

cocreature committed:
Ignore tempfiles, stack.yaml and dist-newstyle

cocreature committed:
Support LLVM-3.9

bscarlet commented:
I've created an llvm-3.9 branch off of llvm-3.8. Apparently github has a new feature to allow changing the base branch of a PR:

https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

but for some reason my new llvm-3.9 branch doesn't show up in the list when I try that for your PR.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Attribute.h:
```diff
@@ -1,9 +1,13 @@
 #define __LLVM_GENERAL_INTERNAL_FFI__ATTRIBUTES__H__
 
+// The last three arguments are flags indicating if this is a
+// parameter attribute, function result attribute or function attribute.
```
Thank you!

It'd be good if this sort of comment could be in a separate PR for ease of merging between branches.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Attribute.h:
```diff
@@ -1,9 +1,13 @@
 	macro(None,F,F,F)                                 \
-	macro(Alignment,T,F,F)                            \
+    macro(Alignment,T,F,F)                            \
+    macro(AllocSize,F,F,T)                            \
```
Whitespace consistency please.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/FunctionC.cpp:
```diff
@@ -55,8 +55,15 @@ void LLVM_General_SetFunctionCallingConvention(LLVMValueRef f, unsigned cc) {
+        f->setPersonalityFn(nullptr);
+    } else {
+        f->setPersonalityFn(unwrap<Constant>(personalityFn));
+    }
```
Might be more readable if you just use a ?: - i.e.

unwrap<Function>(fn)->setPersonalityFn(personalityFn ? unwrap<Constant>(personalityFn) : nullptr);

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/GlobalValue.h:
```diff
@@ -42,6 +42,6 @@ LLVM_GENERAL_FOR_EACH_COMDAT_SELECTION_KIND(ENUM_CASE)
 	macro(LocalDynamicTLSModel)                           \
 	macro(InitialExecTLSModel)                            \
-	macro(LocalExecTLSModel)
+    macro(LocalExecTLSModel)
```
More whitespace consistency.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/GlobalValueC.cpp:
```diff
@@ -40,12 +40,36 @@ void LLVM_General_SetCOMDATSelectionKind(Comdat &comdat, unsigned csk) {
+    None   = 0,
+    Local  = 1,
+    Global = 2,
+} LLVM_General_UnnamedAddr;
```
There's nothing to enforce the non-local correlation between this enum and the corresponding values in the the coding instances made by the call to genCodingInstance you've added to Module.hs. Please follow the example of how Atomicity is done to
- keep correctness as much as possible a consequence of locally verifiable correctness - e.g. lists of pairs where the elements of each pair correspond - as opposed to non-local correlations between files like this.
- minimize boilerplate through the use of metaprogramming - i.e. TH and ENUM_CASE / LLVM_GENERAL_FOR_EACH_*. Circuitous though the metaprogramming is, as much boilerplate as would be necessary without it quickly becomes unmanageable.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/PassManagerC.cpp:
```diff
@@ -85,8 +87,16 @@ void LLVM_General_AddGlobalValueNumberingPass(LLVMPassManagerRef PM, LLVMBool no
 void LLVM_General_AddInternalizePass(LLVMPassManagerRef PM, unsigned nExports, const char **exports) {
-	std::vector<const char *> exportList(exports, exports + nExports);
-	unwrap(PM)->add(createInternalizePass(exportList));
+    const std::vector<const char *> exportList(exports, exports + nExports);
```
No need to allocate/free a vector any more. Just iterate over the array in the lambda.

bscarlet replied:
You'll have copied the array of character pointers, but not the strings the pointers point to. I think you need to use a vector<string> or similar so the lambda owns string data.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/TargetC.cpp:
```diff
@@ -338,22 +316,4 @@ void LLVM_General_InitializeAllTargets() {
   // None of the other components are bound yet
 }
 
-LLVMBool LLVM_General_TargetMachineEmit(
```
What's your motivation for removing this function and changing the factoring of the various emitting functions in Module.hs?

cocreature replied:
This function broke with the 3.9 update and iirc it also already broke with the 3.8 update. I don’t see any good reason to use this function directly and by limiting us to the API provided by the LLVM C API where possible, the amount of breakage to be expected will be smaller.

bscarlet replied:
Please put the previous structure back. The point of exposing the interface which writes to streams is to eventually reduce the number of copies of the output code when doing stuff like JIT or streaming the result over the network.

cocreature committed:
Add new attributes to AST

cocreature committed:
Use tabs instead of spaces

cocreature committed:
Use metaprogramming to generate UnnamedAddr

cocreature committed:
Remove debug output

cocreature committed:
Use a ternary operator in setPersonalityFn

cocreature committed:
Copy strings into lambda

bscarlet commented on llvm-general/src/LLVM/General/Internal/Attribute.hs:
```diff
@@ -50,6 +51,12 @@ instance Monad m => EncodeM m A.PA.ParameterAttribute (Ptr FFI.ParameterAttrBuil
+      A.PA.WriteOnly -> FFI.parameterAttributeKindWriteOnly
+      A.PA.Alignment _ -> error "Impossible"
+      A.PA.Dereferenceable _ -> error "Impossible"
+      A.PA.DereferenceableOrNull _ -> error "Impossible"
```
If you're going spend the code space on these cases, please make their error messages actually provide useful information as to how to fix the presumed problem, like "llvm-general internal error: cases inconsistent in ParameterAttribute encoding for " ++ show a

cocreature committed:
Improve error messages on pattern match failure

cocreature commented:
I think I addressed all your comments. Travis is still failing, haven’t yet figured out why that is. Also I’m still not able to select `llvm-3.9` as a base branch (I was hoping for some cache invalidation issue). Weirdly enough on https://github.com/bscarlet/llvm-general/tree/llvm-3.9 there is a link to this PR. It would be easiest if you just merge manually once you think it’s ready but I’m fine with opening a new PR if you prefer that.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Module.hs:
```diff
@@ -217,17 +219,39 @@ getDataLayout m = do
+  astToFFIModule' mod ffiMod
+  liftIO $ f (Module ffiMod)
+
+astToFFIModule :: Context -> A.Module -> ExceptT String IO Module
```
What is the purpose of factoring withModuleFromAST into different object-management patterns as you have? It is a goal of llvm-general to shield the haskell user from as much having to worry about C++ lifetime management as possible. Where you pushed here by the linking API?

cocreature replied:
Yep, the linking API destroys modules so always calling disposemodule is no longer valid.

bscarlet commented on llvm-general/src/LLVM/General/Module.hs:
```diff
@@ -5,6 +5,7 @@ module LLVM.General.Module (
     File(..),
 
     withModuleFromAST,
+    astToFFIModule,
```
Exposing this function puts the responsibility of understanding how a particular module's memory will be managed into the user - some automatically cleaned up, some not. Linking does pose some challenges, but I'm skeptical this is the right solution.

bscarlet commented on llvm-general/test/LLVM/General/Test/Linking.hs:
```diff
@@ -59,9 +59,9 @@ tests = testGroup "Linking" [
-          moduleAST m0
+      withModuleFromAST' context ast0 $ \dest -> do
+        src <- failInIO $ astToFFIModule context ast1
+        failInIO $ linkModules dest src
```
This test demonstrates my concern. Now users who want to link have to make their "src" module one way, but users who don't or users who do but are making their "dest" should make them another way, and all users have to be mindful of the difference.

A better type for linkModules would express its constraints better. The previous behavior was a compromise. Is the change necessary for 3.9?

cocreature replied:
This change is necessary for 3.9 since LLVM now destroys the src module. Using `withModuleFromAST'` tries to destroy the module which causes segfaults.

I guess we could make a newtype for modules that are consumed but until we get linear types or something like that in Haskell I don’t see how we can actually protect users from reusing the module after at has been destroyed.

bscarlet replied:
Agreed it'd be a pain to handle statically. Can we handle it dynamically? The old version used to cannibalize the module anyway, so it was effectively mutated. How about an approach where the Link function mutates the Module, so the dispose doesn't crash? I.e. change Module from
Module (Ptr FFI.Module) to Module (IORef (Ptr FFI.Module)) and set it to null in the link function, or something like that.

cocreature replied:
I’ve gone for a separate type `ModuleRef` rather than `Module` which is basically your ioref suggestion. The reason for this are a) it makes it explicit in the type what can happen to the module b) we don’t introduce unnecessary overhead where we don’t need it (which is everywhere except link) and c) the only way to make this API safe (for users) is to remove the possiblity to do anything with the module after it has been deleted. By not exposing the `getModule` function the user loses that possibility and it should be (reasonably) safe. Given that the only reason why you would want ModuleRef is to be able to call functions that destroy the module this is a reasonable thing to do. 

However it removes the possibilty to first convert the module and then call some other functions before finally destroying it. I am not sure how important this is, but if you think we need this I would just expose `getModule` and put a warning sign on it.

bscarlet replied:
I don't think this direction with both ModuleRef and Module has worked out very well.
1) Both in fact represent references to mutable modules (just at different levels of indirection), so the distinction makes little sense from the outside.
2) It would in fact be quite common to create a ModuleRef and then want to call functions on the module before consuming it. That's how one would optimize before linking, for example.
3) Exposing getModule just puts us back where we were before this attempt.
4) As written, one can only link modules which come from ASTs, not from LLVMAssembly or bitcode. Solving that problem would in this approach require doubling all the various withModuleFrom... functions
5) You've made this attempt by duplicating functions, which is not ok in it's own right.
6) The overhead of using an IORef is not relevant in this context.

Please just make Module have an IORef.

I looked into the changes in LLVM which drove you here. The important part seems already to be present in 3.8, so perhaps these changes would be better made on that branch.

cocreature replied:
We don’t need this change in 3.8 because we are using the old linking function there (which got removed in 3.9). I am not going to backport this change but feel free to do so yourself.

bscarlet commented on llvm-general/test/LLVM/General/Test/Metadata.hs:
```diff
@@ -73,6 +73,7 @@ tests = testGroup "Metadata" [
           MetadataNodeDefinition (MetadataNodeID 0) [ Just $ MDValue $ ConstantOperand (C.Int 32 1) ]
          ]
     let s = "; ModuleID = '<string>'\n\
+            \source_filename = \"<string>\"\n\
```
If this source filename shows up in serialized modules now, have we provided a way to specify the value?

bscarlet replied:
Has it been added? If not, then perhaps in a following CL?

cocreature replied:
Hm I thought I added it, but apparently I lost the commit somewhere. Let’s move this to a separate PR.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Module.hs:
```diff
@@ -217,17 +219,39 @@ getDataLayout m = do
+  (FFI.unnamedAddrNone, Nothing),
+  (FFI.unnamedAddrLocal, Just A.G.LocalAddr),
+  (FFI.unnamedAddrGlobal, Just A.G.GlobalAddr)
+ ]
```
I think this instance would be better in Global.hs.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Attribute.hs:
```diff
@@ -126,6 +128,25 @@ foreign import ccall unsafe "LLVM_General_AttrBuilderAddAlignment" attrBuilderAd
+attributeGetAllocSizeArgs :: FunctionAttribute -> IO (Word64, Maybe Word64)
+attributeGetAllocSizeArgs attrBuilder =
+  alloca $ \x ->
+  alloca $ \y ->
```
This kind of use of alloca/peek is what the DecodeAST stuff is for. This level of the process belongs up out of the FFI directory, which is as much just-for-imports as I can get it. One exception I've made is currying, where the general Encode/Decode mechanics sometimes work better with partially uncurried types which the imports can't express (e.g. pairing successive arguments into (pointer,length) pairs to pass arrays.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/AttributeC.cpp:
```diff
@@ -113,8 +113,31 @@ void LLVM_General_AttrBuilderAddStackAlignment(AttrBuilder &ab, uint64_t v) {
+    if (optionalIsThere) {
+        ab.addAllocSizeAttr(x, y);
+    } else {
+        ab.addAllocSizeAttr(x, Optional<unsigned>());
```
?:

cocreature commented on llvm-general/test/LLVM/General/Test/Metadata.hs:
```diff
@@ -73,6 +73,7 @@ tests = testGroup "Metadata" [
           MetadataNodeDefinition (MetadataNodeID 0) [ Just $ MDValue $ ConstantOperand (C.Int 32 1) ]
          ]
     let s = "; ModuleID = '<string>'\n\
+            \source_filename = \"<string>\"\n\
```
Good catch, I’ll add that.

bscarlet commented on llvm-general-pure/src/LLVM/General/AST/FunctionAttribute.hs:
```diff
@@ -37,7 +37,14 @@ data FunctionAttribute
         stringAttributeValue :: String -- ^ Use "" for no value -- the two are conflated
-      } 
+      }
+    | AllocSize Word64 (Maybe Word64)
```
You've made these Word64, but they're "unsigned" in the C++ API. They should be Word.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Attribute.hs:
```diff
@@ -58,7 +65,13 @@ instance Monad m => EncodeM m A.FA.FunctionAttribute (Ptr FFI.FunctionAttrBuilde
     A.FA.StackAlignment v -> FFI.attrBuilderAddStackAlignment b v
+    A.FA.AllocSize x y -> case y of
+                            Nothing -> FFI.attrBuilderAddAllocSize b x 0 0
+                            Just y' -> FFI.attrBuilderAddAllocSize b x 1 y'
```
Better to avoid needing this kind of case and the magic use of 0/1 in the third argument.

Use currying/uncurrying helpers next to the FFI import (like constNamedStruct, for example) to bundle the flag and value together into a pair (or newtype with the others in LLVMCTypes like NothingAsMinusOne), and encode/decode instances in Coding.hs so that you can encode/decode y like everything else.

I note I've asked you elsewhere to move the corresponding decode function which you put in the FFI directory down into this directory somewhere. Please try to keep the encode/decode paths parallel.

cocreature replied:
I need a case somewhere, I can either leave it here or push it in the FFI directory but using a tuple doesn’t change a fact that at some point I need to case on the maybe.

bscarlet replied:
Certainly. Haskell isn't that much else, after all. The case would wind up in the encode instance in Coding.hs (where it would be reusable by for any other FFI function which winds up with a similar signature.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Attribute.hs:
```diff
@@ -120,6 +143,9 @@ instance DecodeM DecodeAST A.FA.FunctionAttribute FFI.FunctionAttribute where
          enum <- liftIO $ FFI.functionAttributeKindAsEnum a
          case enum of
+           [functionAttributeKindP|AllocSize|] -> do
+             (x,y) <- liftIO $ FFI.attributeGetAllocSizeArgs a
```
As noted at the definition of attributeGetAllocSizeArgs, it really belongs in this directory. Please try to keep the encode/decode logic parallel.

bscarlet commented:
I think I've now commented on everything I think needs changes. Sorry my comments get stretched out over time. I think this sort of process would go better in future with more, smaller PRs so each can get committed along the way (and so I don't have to re-review so much).

cocreature commented:
> I think this sort of process would go better in future with more, smaller PRs so each can get committed along the way (and so I don't have to re-review so much).

Comitting some intermediate state that doesn’t even compile or fails 90% of the tests is not particularly useful.

cocreature committed:
Expose RawPWriteStream for assembly serialization

cocreature committed:
Move (de)serialization instance of UnnamedAddr

cocreature committed:
Cleanup encoding/decoding of AllocSize

bscarlet commented:
I disagree. The evolution of the code from the initial sad state you describe to all-green is a fact, whether hidden or not. By exposing the steps to get there in more commits & PRs; it's easier to discuss individual changes; it's easier to build consensus on the principles of the codebase code earlier, so later changes involving the same principles can proceed with less friction; and we gain a ratchet - each piece I accept done with, I don't need to re-review it as you make other changes, nor are you in any doubt about it as you make further changes.

cocreature committed:
Hide module destruction from user

bscarlet commented on llvm-general/src/LLVM/General/Internal/Module.hs:
```diff
@@ -80,20 +96,22 @@ genCodingInstance [t| Bool |] ''FFI.LinkerMode [
   (FFI.linkerModeDestroySource, False)
```
I think this enum is unused after the old link function goes away.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Module.hs:
```diff
@@ -80,20 +96,22 @@ genCodingInstance [t| Bool |] ''FFI.LinkerMode [
+  result <- decodeM =<< liftIO (FFI.linkModules dest src)
+  -- linkModules takes care of deleting the sourcemodule
+  liftIO $ deleteModule srcRef
+  when result (throwError "Couldn’t link modules")
```
Oh yuck. The real messages have been shuffled into state hanging off the context. It's pretty ugly to have a signature promising error messages and then deliver only a made up "didn't work" string.

We could expose the diagnostic handler in the context and give this function a simpler signature. Significantly better, I think, would be to implement the diagnostic handler ourselves and return the actual errors here.

cocreature replied:
Agreed, but please let us do that in a separate PR.

cocreature committed:
Remove separate ModuleRef type

cocreature committed:
Remove LinkerMode enum

cocreature commented:
I think I addressed all your comments.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Analysis.hs:
```diff
@@ -16,8 +16,9 @@ import LLVM.General.Internal.Coding
   errorPtr <- alloca
-  result <- decodeM =<< (liftIO $ FFI.verifyModule m FFI.verifierFailureActionReturnStatus errorPtr)
+  m' <- readModule m
+  result <- decodeM =<< (liftIO $ FFI.verifyModule m' FFI.verifierFailureActionReturnStatus errorPtr)
```
Not strictly necessary, but if you put the prime on the first m (or neither) instead of the second, then you won't need a change and diffs will be smaller.

cocreature replied:
I deliberately use a prime here and putting it on the second is more confusing imho.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Attribute.hs:
```diff
@@ -50,15 +51,27 @@ instance Monad m => EncodeM m A.PA.ParameterAttribute (Ptr FFI.ParameterAttrBuil
+      A.PA.WriteOnly -> FFI.parameterAttributeKindWriteOnly
+      A.PA.Alignment _ -> error $ "llvm-general internal error: cases inconsistent in ParameterAttribute encoding for " ++ show a
+      A.PA.Dereferenceable _ -> error $ "llvm-general internal error: cases inconsistent in ParameterAttribute encoding for " ++ show a
+      A.PA.DereferenceableOrNull _ -> error $ "llvm-general internal error: cases inconsistent in ParameterAttribute encoding for " ++ show a
```
Please factor these long error messages into a helper function (shared with the other cases, below).

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Attribute.hs:
```diff
@@ -126,6 +126,18 @@ foreign import ccall unsafe "LLVM_General_AttrBuilderAddAlignment" attrBuilderAd
+  Ptr FunctionAttrBuilder -> CUInt -> CUInt -> LLVMBool -> IO ()
+
+attrBuilderAddAllocSize :: Ptr FunctionAttrBuilder -> CUInt -> (CUInt, LLVMBool) -> IO ()
+attrBuilderAddAllocSize b i (y,isJust) = attrBuilderAddAllocSize' b i y isJust
```
space after commas

bscarlet commented on llvm-general/src/LLVM/General/Internal/Attribute.hs:
```diff
@@ -50,15 +51,27 @@ instance Monad m => EncodeM m A.PA.ParameterAttribute (Ptr FFI.ParameterAttrBuil
+    A.FA.StackAlignment v -> liftIO $ FFI.attrBuilderAddStackAlignment b v
+    A.FA.AllocSize x y -> do x' <- encodeM x
+                             y' <- encodeM y
+                             liftIO $ FFI.attrBuilderAddAllocSize b x' y'
```
formatting:
multi-line do-blocks wrapped after the "do", indented with as few units of 2 spaces as still works.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Attribute.hs:
```diff
@@ -120,6 +143,14 @@ instance DecodeM DecodeAST A.FA.FunctionAttribute FFI.FunctionAttribute where
+             isJust <- liftIO $ FFI.attributeGetAllocSizeArgs a x y
+             x' <- decodeM =<< peek x
+             y' <- peek y
+             yM <- decodeM (y',isJust)
```
formatting: space after comma

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Attribute.h:
```diff
@@ -6,9 +6,10 @@
+	macro(AllocSize,F,F,T)                            \
 	macro(AlwaysInline,F,F,T)                         \
-	macro(ArgMemOnly,T,F,T)                           \
+	macro(ArgMemOnly,F,F,T)                           \
```
Did this change llvm-3.8 to llvm-3.9, or is this a fix that belongs there too?

cocreature replied:
A general fix, I extracted it in a separate commit to make it easier to cherry pick.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Instruction.h:
```diff
@@ -4,7 +4,7 @@
 
 #define LLVM_GENERAL_FOR_EACH_ATOMIC_ORDERING(macro) \
-	macro(NotAtomic) \
+    macro(NotAtomic)                 \
```
Is this change intentional? In github it looks like it's just messing up the formatting.

cocreature replied:
No sorry, I messed this up.

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/Target.h:
```diff
@@ -34,8 +34,7 @@
 	macro(GuaranteedTailCallOpt)													\
-	macro(EnableFastISel)																	\
-	macro(PositionIndependentExecutable)									\
+    macro(EnableFastISel)																	\
```
indent doesn't look consistent

bscarlet commented on llvm-general/src/LLVM/General/Internal/FFI/TargetC.cpp:
```diff
@@ -338,22 +316,42 @@ void LLVM_General_InitializeAllTargets() {
+  }
+  if (TM->addPassesToEmitFile(pass, OS, ft)) {
+    error = "TargetMachine can't emit a file of this type";
+    *ErrorMessage = strdup(error.c_str());
```
Why the use of the std::string error? It looks like it just adds an allocation/deallocation.

cocreature replied:
This implementation is exactly identical to the one in LLVM (as I mentioned in the comment). I don’t see any good reason to diverge from this as it will only mean more effort when updating and we have to figure out why something broke.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Module.hs:
```diff
@@ -217,17 +236,30 @@ getDataLayout m = do
+  moduleId <- encodeM (A.moduleName mod)
+  m <- anyContToM $ bracket (newModule =<< FFI.moduleCreateWithNameInContext moduleId c) (FFI.disposeModule <=< readModule)
+  m' <- readModule m
+  astToFFIModule mod m'
```
Please inline astToFFIModule back into withModuleFromAST, as it was originally.

bscarlet commented on llvm-general/src/LLVM/General/Internal/Module.hs:
```diff
@@ -347,25 +379,23 @@ withModuleFromAST context@(Context c) (A.Module moduleId dataLayout triple defin
 moduleAST :: Module -> IO A.Module
-moduleAST (Module mod) = runDecodeAST $ do
-  c <- return Context `ap` liftIO (FFI.getModuleContext mod)
+moduleAST mod = runDecodeAST $ do
```
Please readModule mod once here, rather than at each use of mod, below.

bscarlet commented:
Just style nits left. After this, looks good. I see you've changed the base branch - github support let me know they've fixed the bug that kept llvm-3.9 from showing up.

cocreature committed:
Extract error messages in a separate function

cocreature committed:
Style fixes

cocreature committed:
Only call readModule once in moduleAST

cocreature committed:
Inline astToFFIModule

cocreature committed:
Set ArgMemOnly parameterflag to false

cocreature commented:
> Just style nits left. After this, looks good.

Show all be fixed now

> I see you've changed the base branch - github support let me know they've fixed the bug that kept llvm-3.9 from showing up.

Yeah I contacted github support as well :)

bscarlet commented:
See issue #185 for some details of the travis failure.
